### ISSUE TITLE
Troubles with thread affinity

### ISSUE THREAD

**Issue by gweodoo** [dmtcp/dmtcp#605]
Hi guys,

Still on the way to integrate DMTCP inside an MPI implementation, I'm facing an issue with thread affinity.
If I set each thread on a dedicated core (through sched_setaffinity, pthread_setaffinity_np, etc...), everything goes well. I can trigger some checkpoints without affecting thread pinning policy.

The issue arises when the program attempts to be restarted. Threads are not bound anymore, and own the entire cpuset !
I found a test (`sched_test.c`) having this case and I was impressed it worked. When un-commenting `assert(CPU_ISSET(i, &cset))` and adding `CPU_COUNT(&cset) == 1` to the condition, the restart effectively failed. This is because, at restart time, each thread inherits the complete topology. As the test evaluates each CPU value independently, all conditions will be true.

Do you know where this could be come from ? It thread affinity, part of DMTCP's job ?
I'm still working with the 2.5 branch + GCC 6.2.0.
At the end, you will find my sample test case, based on hwloc for portability, if needed.

Thanks for your help.

```c
#define _GNU_SOURCE
#include <hwloc.h>
#include <sched.h>
#include <pthread.h>

hwloc_topology_t topo;

void* thread(void * arg)
{
	char* buf = NULL;

	hwloc_cpuset_t set = hwloc_bitmap_alloc();
	hwloc_bitmap_zero(set);
	
	hwloc_get_thread_cpubind(topo, pthread_self(), set, HWLOC_CPUBIND_THREAD);
	hwloc_bitmap_list_asprintf(&buf, set);
	fprintf(stderr, "1. cpuset = %s\n", buf);
	free(buf);

	sleep(5);

	hwloc_bitmap_zero(set);
	hwloc_get_thread_cpubind(topo, pthread_self(), set, HWLOC_CPUBIND_THREAD);
	hwloc_bitmap_list_asprintf(&buf, set);
	fprintf(stderr, "2. cpuset = %s\n", buf);
	free(buf);
	return NULL;
}

int main(int argc, char *argv[])
{
	hwloc_topology_init(&topo);
	hwloc_topology_load(topo);

	int i = 0;
	const int max = 4;
	pthread_t tt[max];
	for(i= 0; i < max; i++)
	{
		pthread_attr_t attr;
		hwloc_obj_t obj = hwloc_get_obj_by_type(topo, HWLOC_OBJ_CORE, i);
		pthread_attr_init(&attr);

		cpu_set_t cpuset;
		CPU_ZERO(&cpuset);
		CPU_SET(hwloc_bitmap_first(obj->cpuset), &cpuset);
		pthread_create(tt+i, &attr, thread, NULL);
		pthread_attr_destroy(&attr);
	}

	for(i= 0; i < max; i++)
		pthread_join(*(tt +i), NULL);

	hwloc_topology_destroy(topo);
	return 0;
}
```

**Comment by gweodoo** [comment]
--oops: Maybe it is related to #203 ?

**Comment by rohgarg** [comment]
Sorry about the delay, @gweodoo. Thanks for reporting this. This certainly looks related to #203. We haven't had time in the past to work on this, but this is not hard to do using a plugin. I'll try to write something and hopefully get back to you soon.

**Comment by gweodoo** [comment]
Thanks for the reply @rohgarg !

It would probably easy to register for each thread a CPU id, and set the affinity at post-{checkpoint,restart}. the NPTL can clearly be used for this.

But, from a portable point of view, it should be preferable for DMTCP to use a dedicated tool like ([hwloc](https://www.open-mpi.org/projects/hwloc/), widely used in HPC community (but maybe you already knew it). It could be embedded in DMTCP as a dep by default and installed during the DMTCP building process (hwloc is under BSD license)

**Comment by rohgarg** [comment]
> It would probably easy to register for each thread a CPU id, and set the affinity at post-{checkpoint,restart}. the NPTL can clearly be used for this.

Yes, that's what I had in mind.

> But, from a portable point of view, it should be preferable for DMTCP to use a dedicated tool like (hwloc, widely used in HPC community (but maybe you already knew it). It could be embedded in DMTCP as a dep by default and installed during the DMTCP building process (hwloc is under BSD license)

Actually, I'm not that familiar with the hwloc API, so I wasn't really planning on using it here. My plan was to use the POSIX/Linux calls to save and restore affinities, at least to demonstrate a proof-of-concept. But I think this is a good point.

That said, since DMTCP is Linux-specific (at least, right now), I am not sure if we should really be worried about portability.

In any case, I think if we choose to go the hwloc route, we don't really have to add a hard dependency on hwloc for core DMTCP. My reasoning is that it could end up breaking things for many non-HPC and HPC users of DMTCP. This could be a separate plugin that gets compiled optionally.

Please let me know what you think.

(Out of curiosity, which MPI implementation are you working with? Also, would it be easy for you to work with the 3.0 (master) branch? I'm not too keen on adding new things to the 2.5 branch.)

**Comment by gweodoo** [comment]
> That said, since DMTCP is Linux-specific (at least, right now), I am not sure if we should really be worried about portability.

You are probably right, no need to make things more complicated. DMTCP is not a HPC-specific tool at first. (Almost) every user relies on the NPTL implementation.

I'm currently working on a MPI implementation called MPC (link [here](http://mpc.hpcframework.paratools.com)) which, basically, aims to encapsulate MPI "processes" inside its own user-level threads. Especially, I'm studying how this implementation could rely on DMTCP as a C/R system. While I am on the subject, If you are interested, it could be productive to discuss about DMTCP for HPC needs (particularly concerning performance), instead of opening a ticket each time. Why not providing you a guest account to help to maintain DMTCP for HPC ?

Some time ago, I tried the 3.0 branch but faced some issues to have a working case. I think that it was due to my own environment configuration, I will probably re-try it soon.

**Comment by rohgarg** [comment]
> Why not providing you a guest account to help to maintain DMTCP for HPC ?

Sure, we'd be happy to collaborate on and support different use-cases. You can write to the DMTCP dev list or to me directly.

**Comment by rohgarg** [comment]
@gweodoo: Could you please check if PR #606 works for you? (It is based on the 2.5 branch.)

**Comment by gweodoo** [comment]
Yes, thanks for the quickfix ! I will try to check out the 3.0 branch instead

**Comment by rohgarg** [comment]
I just realized that the 3.0 branch (master) will need a few more changes. The DMTCP events that I'm using in PR #606 have been deleted from the 3.0 branch.  We will need to bring them back to get this to work.
